### PR TITLE
[SPARK-2883][SQL] Spark Support for ORCFile format

### PR DESCRIPTION
**Pull request by zhzhan**
This is the ORC support for new external data source API. 

@scwf and I have collaborated on this PR for a long time and have patches reviewed. The major piece is done long time before, but the PR got delayed due to restructure of spark data source API.  @scwf also has a PR under review, which seems to be inactive.  Anyway, this is the work done by both @scwf and me.

The current patch is independent from any other spark package, and resides in spark/hive project, due to the dependency on orc file format. 

Major feature:
1. New data source API support.
2. Basic Operator: saveAsOrcFile and OrcFile. The former is used to save the table into orc format file, and the latter is used to import orc format file into spark sql table.
3. Column pruning
4. Self-contained schema support: The orc support is fully functional independent of hive metastore. The table schema is maintained by the orc file itself.
5. To support the orc file, user need to: import import org.apache.spark.sql.hive.orc._ to bring in the orc support into context
6. The orc file is operated in HiveContext, the only reason is due to package issue, and we don’t want to bring in hive dependency into spark sql. Note that orc operations does not relies on Hive metastore.
7. It support full complicated dataType in Spark Sql, for example, list, seq, and nested datatype.

Hive 0.13.1 support.
1. the orc can support different compression method, e.g., SNAPPY, LZO, ZLIB, and NONE
2. prediction pushdown

### CODE REVIEW

**Comment by zhzhan** [comment]
@liancheng and @marmbrus Could you help to review the patch.

**Comment by climberus** [comment]
@zhzhan Do u have the example code to use that ORCContext?  Thanks

**Comment by zhzhan** [comment]
@climberus following examples demonstrate how to use it:

import org.apache.spark.sql.hive.orc._
import org.apache.spark.sql._
//saveAsOrcFile
case class AllDataTypes(
    stringField: String,
    intField: Int,
    longField: Long,
    floatField: Float,
    doubleField: Double,
    shortField: Short,
    byteField: Byte,
    booleanField: Boolean)

    val range = (0 to 255)
    val data = sc.parallelize(range).map(x => AllDataTypes(s"$x", x, x.toLong, x.toFloat, x.toDouble, x.toShort, x.toByte, x % 2 == 0))
    data.toDF().saveAsOrcFile("orcTest")
//read orcFile
val hiveContext = new org.apache.spark.sql.hive.HiveContext(sc)
val orcTest = hiveContext.orcFile("orcTest")
orcTest.registerTempTable("orcTest")
hiveContext.sql("SELECT * from orcTest where intfield>185").collect.foreach(println)

  hiveContext.sql("create temporary table orc using org.apache.spark.sql.hive.orc OPTIONS (path \"orcTest\")")
  hiveContext.sql("select * from orc").collect.foreach(println)
val table = hiveContext.sql("select * from orc")
table.saveAsTable("table", "org.apache.spark.sql.hive.orc")
val hiveOrc = hiveContext.orcFile("/user/hive/warehouse/table")
hiveOrc.registerTempTable("hiveOrc")
hiveContext.sql("select * from hiveOrc").collect.foreach(println)
table.saveAsOrcFile("/user/ambari-qa/table")
hiveContext.sql("create temporary table normal_orc_as_source USING org.apache.spark.sql.hive.orc OPTIONS (path 'saveTable') as select * from table")

**Comment by climberus** [comment]
awesome

**Comment by liancheng** [comment]
@zhzhan I'm right now designing partitioning support for the data sources API, and will hopefully make the design doc next week. Will come back to this PR after that. With that part at hand, I believe we can further simplify the ORC data source.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/package.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      val parameters = Map("path"->path)
+      val relation = OrcRelation(path, parameters, Some(dataFrame.schema))(dataFrame.sqlContext)
+      relation.insert(dataFrame, false)
+    }
+  }
+
+  // for orc compression type, only take effect in hive 0.13.1
+  val orcDefaultCompressVar = "hive.exec.orc.default.compress"
+  // for prediction push down in hive-0.13.1, don't enable it
+  var ORC_FILTER_PUSHDOWN_ENABLED = true
```
Comment (deanchen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/package.scala]: How can this variable be disabled?

Comment (zhzhan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/package.scala]: @deanchen You can disable it by 
spark.sql.hive.orc.ORC_FILTER_PUSHDOWN_ENABLED = false

**Comment by zhzhan** [comment]
Will close this with new partition API support.

**Comment by richardxin** [comment]
what will be Java equivalent for import org.apache.spark.sql.hive.orc._? Which dependency I need to put in pom.xml?  I am not able to find dependency for maven pom.xml.

**Pull request closed without merging**